## Analisi Anagrafe Spafid

Al fine di verificare la possibilità di utilizzare i flussi anagrafici forniti da Spafid nell'ambito del progetto di automatizzazione del processo delle __"Schede Operazione"__, è stato necessario procedere con un'attività di analisi preventiva sui dati forniti.
I flussi sono i seguenti:
1. `Stampa Elenco Rapporti.xlsx`
2. `Stampa Elenco Clienti.xlsx`

Tale analisi è consistita nel:
- verificare la corrispondenza di ogni rapporto coi relativi clienti
- ......

Al fine di consentire l'importazione in pandas di files excel è stato necessario installare xldr:  
`> conda install xlrd`



In [2]:
import pandas as pd

data = pd.read_excel (r'Stampa Elenco Rapporti.xlsx', sheet_name='lista', skiprows=1)
df_rapporti = pd.DataFrame(data, columns= ['Codice rapporto','Tipo mandato', 'Data apertura', 'Rif. interno'])
df_rapporti

Codice rapporto                               Tipo mandato Data apertura  \
0           00300266      00021 AMMINISTRAZIONE TITOLI E VALORI    2017-07-10   
1           IFP09001            IFID5 MANDATO FIDUCIARIO TITOLI    2014-08-01   
2           00300916        00031 MANDATO FIDUCIARIO PER ESCROW    2020-03-27   
3           00300676  00030 MANDATO FIDUCIARIO PER STOCK-OPTION    2019-06-05   
4           00300923        00031 MANDATO FIDUCIARIO PER ESCROW    2020-03-31   
...              ...                                        ...           ...   
2162        IFO00606                  IFID1 FIDUCIARIO CLASSICO    2014-08-01   
2163        ES000960                00001 FIDUCIARIO SOCIETARIO    2018-06-30   
2164        00300657        00031 MANDATO FIDUCIARIO PER ESCROW    2019-05-07   
2165        IFC19101                  IFID1 FIDUCIARIO CLASSICO    2014-08-01   
2166        00300621      00021 AMMINISTRAZIONE TITOLI E VALORI    2019-04-16   

               Rif. interno  
0     CB BARBAGLIO CAROLINA  
1           PA PAVAN ANDREA  
2     CB BARBAGLIO CAROLINA  
3     CB BARBAGLIO CAROLINA  
4     CB BARBAGLIO CAROLINA  
...                     ...  
2162        PA PAVAN ANDREA  
2163      AC CHIESA ALBERTO  
2164      AC CHIESA ALBERTO  
2165        PA PAVAN ANDREA  
2166      AC CHIESA ALBERTO  

[2167 rows x 4 columns]

In [3]:
data = pd.read_excel (r'Stampa Elenco Clienti.xlsx', sheet_name='lista', skiprows=1)
df_clienti = pd.DataFrame(data, columns= ['Codice cliente','Rapporto', 'Tipo Mandato', 'Tipo legame', 'Data apertura'])
df_clienti

Codice cliente  Rapporto  \
0             A0180  00122900   
1             A0690  00300855   
2             A0900  00184417   
3             A0940  00184440   
4             A0950  00198123   
...             ...       ...   
4994          10055  00300025   
4995          10056  00300026   
4996          10057  00300026   
4997          10058  00300985   
4998          10058  00300986   

                                           Tipo Mandato  \
0       00007 - FIDUCIARIO SOCIETARIO SENZA COMMISSIONI   
1           00030 - MANDATO FIDUCIARIO PER STOCK-OPTION   
2                         00001 - FIDUCIARIO SOCIETARIO   
3       00007 - FIDUCIARIO SOCIETARIO SENZA COMMISSIONI   
4                00004 - FIDUCIARIO TITOLI  POSTICIPATO   
...                                                 ...   
4994              00031 - MANDATO FIDUCIARIO PER ESCROW   
4995              00031 - MANDATO FIDUCIARIO PER ESCROW   
4996              00031 - MANDATO FIDUCIARIO PER ESCROW   
4997  00026 - AMM. TITOLI, VALORI, PART. SOCIETA', A...   
4998  00026 - AMM. TITOLI, VALORI, PART. SOCIETA', A...   

                 Tipo legame Data apertura  
0          A - Cliente unico    15/06/1988  
1          A - Cliente unico    21/02/2020  
2            P - Procuratore    10/03/2011  
3          A - Cliente unico    24/05/2004  
4          A - Cliente unico    24/04/2014  
...                      ...           ...  
4994       A - Cliente unico    27/06/2016  
4995       A - Cliente unico    27/06/2016  
4996  X - Titolare effettivo    27/06/2016  
4997  X - Titolare effettivo    10/07/2020  
4998  X - Titolare effettivo    10/07/2020  

[4999 rows x 5 columns]

Partendo dai 2 dataframe appena creati è possibile effettuarne il merge utilizzando come campo di join la colonna `Rapporto`.
An inner join combines two DataFrames based on a join key and returns a new DataFrame that contains only those rows that have matching values in both of the original DataFrames.
Vedi https://datacarpentry.org/python-ecology-lesson/05-merging-data/

In tal modo, partendo dall'elenco dei clienti, è possibile associare a ciascun rapporto il rispettivo Rif. interno (non serve a nulla ma è un esercizio) :-).


In [35]:
merged_inner = pd.merge(left=df_clienti, right=df_rapporti, left_on='Rapporto', right_on='Codice rapporto')
merged_inner
merged_inner.reset_index(drop=True)

Codice cliente  Rapporto  \
0             A0180  00122900   
1             M0360  00122900   
2             S1900  00122900   
3             A0690  00300855   
4          00010673  00300855   
...             ...       ...   
4994          10045  00300027   
4995          10046  00300027   
4996          10047  00300438   
4997          10049  00300020   
4998          10050  00300020   

                                         Tipo Mandato  \
0     00007 - FIDUCIARIO SOCIETARIO SENZA COMMISSIONI   
1     00007 - FIDUCIARIO SOCIETARIO SENZA COMMISSIONI   
2     00007 - FIDUCIARIO SOCIETARIO SENZA COMMISSIONI   
3         00030 - MANDATO FIDUCIARIO PER STOCK-OPTION   
4         00030 - MANDATO FIDUCIARIO PER STOCK-OPTION   
...                                               ...   
4994             00003 - FIDUCIARIO TITOLI ANTICIPATO   
4995             00003 - FIDUCIARIO TITOLI ANTICIPATO   
4996      00030 - MANDATO FIDUCIARIO PER STOCK-OPTION   
4997             00003 - FIDUCIARIO TITOLI ANTICIPATO   
4998             00003 - FIDUCIARIO TITOLI ANTICIPATO   

                    Tipo legame Data apertura_x Codice rapporto  \
0             A - Cliente unico      15/06/1988        00122900   
1               P - Procuratore      05/10/1998        00122900   
2               C - Controparte      01/01/2016        00122900   
3             A - Cliente unico      21/02/2020        00300855   
4        X - Titolare effettivo      21/02/2020        00300855   
...                         ...             ...             ...   
4994       B - Cliente primario      29/06/2016        00300027   
4995  M - Cliente contestatario      29/06/2016        00300027   
4996          A - Cliente unico      15/11/2017        00300438   
4997       B - Cliente primario      29/06/2016        00300020   
4998  M - Cliente contestatario      29/06/2016        00300020   

                                       Tipo mandato Data apertura_y  \
0     00007 FIDUCIARIO SOCIETARIO SENZA COMMISSIONI      1988-06-15   
1     00007 FIDUCIARIO SOCIETARIO SENZA COMMISSIONI      1988-06-15   
2     00007 FIDUCIARIO SOCIETARIO SENZA COMMISSIONI      1988-06-15   
3         00030 MANDATO FIDUCIARIO PER STOCK-OPTION      2020-02-21   
4         00030 MANDATO FIDUCIARIO PER STOCK-OPTION      2020-02-21   
...                                             ...             ...   
4994             00003 FIDUCIARIO TITOLI ANTICIPATO      2016-06-29   
4995             00003 FIDUCIARIO TITOLI ANTICIPATO      2016-06-29   
4996      00030 MANDATO FIDUCIARIO PER STOCK-OPTION      2017-11-15   
4997             00003 FIDUCIARIO TITOLI ANTICIPATO      2016-06-29   
4998             00003 FIDUCIARIO TITOLI ANTICIPATO      2016-06-29   

                Rif. interno  
0         LF FUMAGALLI LUIGI  
1         LF FUMAGALLI LUIGI  
2         LF FUMAGALLI LUIGI  
3      CB BARBAGLIO CAROLINA  
4      CB BARBAGLIO CAROLINA  
...                      ...  
4994  GM MARTINELLI GIANLUCA  
4995  GM MARTINELLI GIANLUCA  
4996      LF FUMAGALLI LUIGI  
4997  GM MARTINELLI GIANLUCA  
4998  GM MARTINELLI GIANLUCA  

[4999 rows x 9 columns]

Dal momento che una inner join estrae solo i record per i quali la colonna di join risulta presenti in entrambi i df, il fatto che i record estratti siano 4999 (corrispondente al numero di record del df dei clienti), significa che tutti i codici rapporto presenti nel df dei clienti sono presenti nel df dei rapporti. In questo caso il df dei rapporti è una tabella di lookup.


In alternativa posso raggruppare il df dei clienti per la colonna Codice Rapporto e contare quante volte un singolo rapporto è presente nel df dei clienti. 
In tal modo è possibile sapere quanti clienti sono associati a ciascun rapporto.
Per fare ciò sono possibili 2 soluzioni (che ovviamente forniscono lo stesso risultato) :-)

In [38]:
#df_clienti.groupby('Rapporto')
d1 = df_clienti.Rapporto.value_counts().rename_axis('Rapporto').reset_index(name='count')
d1.reset_index(drop=True)
d1.sort_values(by=['Rapporto'], inplace=True)


#In alternativa è possibile utilizzare use groupby + agg:
d2 = df_clienti.groupby('Rapporto').Rapporto.agg([('count', 'count')]).reset_index()
d2.reset_index(drop=True)
d2.sort_values(by=['Rapporto'], inplace=True)


## Primo confronto manuale della corrispondenza del numero dei clienti a parità di rapporto
dd = d1.loc[d1['Rapporto'] == '00300020']
d1

Rapporto  count
582   00000100      3
579   00020600      3
905   00062602      2
234   00079300      4
1006  00079605      2
...        ...    ...
776   VRS06901      2
1231  VRT01204      2
204   VRT01301      5
1241  VRT03801      2
1141  VRZ02001      2

[2167 rows x 2 columns]

Verificare la corrispondenza tra il numero di clienti x rapporto nei 2 df (d1 e d2)

In [28]:

d1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2167 entries, 582 to 1141
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Rapporto  2167 non-null   object
 1   count     2167 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 50.8+ KB


In [29]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2167 entries, 0 to 2166
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Rapporto  2167 non-null   object
 1   count     2167 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 50.8+ KB


una volta verificata l'eguaglianza tra i 2 dataframe (ancora da dimostrare!!!!), possiamo concentrarsi su uno di essi e procedere con l'analisi.
Come primo step ordineremo il df per il numero di clienti x rapporto

In [44]:
d1.sort_values(by=['count'], inplace=True, ascending=False)
d1

Rapporto  count
0     00300728     19
1     00300906     17
3     00300760     14
2     00300329     14
4     00300702     13
...        ...    ...
1984  00300981      1
1924  00300982      1
2097  00300984      1
1405  ESQ_1825      1
1756  00300737      1

[2167 rows x 2 columns]

Il rapporto 00300728, con 19 intestatari, è il rapporto con più clienti associati
A questo punto potremmo estrarre un grafico per rappresentare la distribuzione di frequenza delle cointestazione dei rapporti